src/lib/pubkey/xmss/xmss_publickey.h

```diff
@@ -0,0 +1,273 @@
+       * method.
+       *
+       * @return XMSS parameters.
+       **/
+      const XMSS_Parameters& xmss_parameters() const
+         {
+         return m_xmss_params;
+         }
+
+      /**
+       * Retrieves the Winternitz One Time Signature (WOTS) method,
+       * corrseponding to the chosen XMSS signature method.
+       *
+       * @return XMSS WOTS signature method identifier.
+       **/
+      XMSS_WOTS_Parameters::ots_algorithm_t wots_oid() const
+         {
+         return m_wots_params.oid();
+         }
+
+      /**
+       * Retrieves the Winternitz One Time Signature (WOTS) parameters
+       * corrseponding to the chosen XMSS signature method.
+       *
+       * @return XMSS WOTS signature method parameters.
+       **/
+      const XMSS_WOTS_Parameters& wots_parameters() const
+         {
+         return m_wots_params;
+         }
+
+      secure_vector<byte>& root()
+         {
+         return m_root;
+         }
+
+      void set_root(const secure_vector<byte>& root)
+         {
+         m_root = root;
+         }
+
+      void set_root(secure_vector<byte>&& root)
+         {
+         m_root = std::move(root);
+         }
+
+      const secure_vector<byte>& root() const
+         {
+         return m_root;
+         }
+
+      virtual secure_vector<byte>& public_seed()
+         {
+         return m_public_seed;
+         }
+
+      virtual void set_public_seed(const secure_vector<byte>& public_seed)
+         {
+         m_public_seed = public_seed;
+         }
+
+      virtual void set_public_seed(secure_vector<byte>&& public_seed)
+         {
+         m_public_seed = std::move(public_seed);
+         }
+
+      virtual const secure_vector<byte>& public_seed() const
+         {
+         return m_public_seed;
+         }
+
+      std::string algo_name() const override
+         {
+         return "XMSS";
+         }
+
+      virtual AlgorithmIdentifier algorithm_identifier() const override
+         {
+         return AlgorithmIdentifier(get_oid(), AlgorithmIdentifier::USE_NULL_PARAM);
+         }
+
+      virtual bool check_key(RandomNumberGenerator&, bool) const override
+         {
+         BOTAN_ASSERT(false, "No key strength check implemented for XMSS.");
+         }
+
+      virtual std::unique_ptr<PK_Ops::Verification>
+         create_verification_op(const std::string&,
+                                const std::string& provider) const override;
+
+      virtual size_t estimated_strength() const override
+         {
+         return m_xmss_params.estimated_strength();
+         }
+
+      virtual size_t max_input_bits() const override
+         {
+         return std::numeric_limits<size_t>::infinity();
+         }
+
+      virtual size_t message_part_size() const override
+         {
+         return std::numeric_limits<size_t>::infinity();
+         }
+
+      virtual size_t message_parts() const override
+         {
+         return std::numeric_limits<size_t>::infinity();
```

Comment:
`max_input_bits` is only an implementation function used by `PK_Ops::{Signature,Verification}_with_EMSA` and never called through the PK_Ops::Signature API. I will move the declaration to the sub-class. XMSS is correct to avoid these subclasses since XMSS does message hashing in an unusual way and does not rely on any external padding scheme.

As to the other functions, these have nothing to do with limitations on `update` (no such limits exist beyond what fits in size_t), instead the (confusingly named) `message_part_size` and `message_parts` should only be overridden if the output signature is a set of BigInts and you want to DER encode them (parts=2, size=order bits). This DER encoding is only used for DSA and ECDSA signatures in some protocols. XMSS should simply not define the functions and assume the default.